### ISSUE TITLE
false error with bi directional type inference of type variable resolution

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#1229]
### Summary

```py
from __future__ import annotations

class A[T]:
    def __init__(self, value: T):
        self.t: T = value
    def f[S](self, value: A[S | T], /) -> A[S | T]:
        return A(value.t)


_: A[object] = A(1).f(A(2))  # Argument to bound method `f` is incorrect: Expected `A[object]`, found `A[int]`
```
`A(2)` should be inferred as `A[object]`, not `A[int]`

**Comment by ibraheemdev** [comment]
This was fixed by https://github.com/astral-sh/ruff/pull/21210 and no longer errors on main.
